### ISSUE TITLE
[jvm] Exception on casting Float to Signle

### ISSUE THREAD

**Issue by RealyUniqueName** [HaxeFoundation/haxe#8768]
```haxe
class Main {
	static function foo(?e:Float):Null<Single> {
		return (e:Float);
	}
	static function main() {
		trace( foo() );
	}
}
```
```
$ java -jar Main.jar 
Error: Unable to initialize main class haxe.root.Main
Caused by: java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    haxe/root/Main.foo(Ljava/lang/Double;)Ljava/lang/Float; @7: areturn
  Reason:
    Type 'java/lang/Double' (current frame, stack[0]) is not assignable to 'java/lang/Float' (from method signature)
  Current Frame:
    bci: @7
    flags: { }
    locals: { 'java/lang/Double' }
    stack: { 'java/lang/Double' }
  Bytecode:
    0000000: 2ab8 000a b800 10b0  
```

**Comment by Simn** [comment]
I noted that in #8764 already, but sure I'll take this issue instead.

**Comment by Simn** [comment]
Wait, why is `Float` assignable to `Single`? That's a narrowing conversion which shouldn't be allowed implicitly...

**Comment by kLabz** [comment]
Because `abstract Single to Float from Float`? Or do you mean this abstract shouldn't have `from Float`? =/

**Comment by Simn** [comment]
We can `to Float` but we can't `from Float`. I'll have to support this regardless because someone could just `cast` it, but this shouldn't be allowed by the compiler implicitly.

**Comment by ncannasse** [comment]
We should discuss these at some point, I would like to have UI8 and UI16 as
well, and the rules for automatic conversion are yet to be defined.
For another example we allow Int -> Single while there might be some less
of precision

**Comment by ncannasse** [comment]
There's also the question of intermediate overflows, such as  (132:UI8) *
(5:UI8)

**Comment by Simn** [comment]
With known sizes, the implicit conversion chain is quite obvious:

```
                Single -> Double
                   ^        ^
Int8 -> Int16 -> Int32 -> Int64
```

The only real issue is `Int` because we don't define its size. We could just assume Int64 and disallow the assignment to Single. Nobody uses that anyway, and if they do they can just `cast` it.

Overflow behavior depends on how the binops are typed. If Int8 * Int8 is Int8 then it has a defined overflow behavior.

**Comment by Aurel300** [comment]
I agree with the int conversions, but I don't think the int to float conversions make much sense. Why not allow `Int64` to `Single`? Sure there are fewer bits but they are not directly comparable. The user might care about the numeric range more.

**Comment by Simn** [comment]
Hmm, I thought there was no opcode for this in the JVM, but l2f is indeed a thing. In that case I agree we can just allow it.

**Comment by Simn** [comment]
By the way, what I'm more interested in than implicit conversions is to what extent we have to allow conversions in general, e.g. if `cast`:

* How should  a target behave if someone assigns a `Float` to an `Int8`?
* Even worse, what about `Null<Float>` to `Null<Int8>`?
* Even worse, what about `Null<Float>` hidden behind `Dynamic` to `Null<Int8>`?
* Even worse, what about `Null<Float>` hidden behind `Dynamic` to `Null<Int8>` hidden behind `Dynamic`?
* Even worse, ... no I don't think it could get worse.

**Comment by ncannasse** [comment]
It's not that obvious : you are losing some bits when doing Int32 > Single or Int64 > Double for instance.

So either we go with "strict" behavior, where any loss of data requires a manual cast, or we go with a "loose" one where every conversion is automatic.

And there is two kind of casts : 
- "automatic" casts, with Dynamic-to-X conversion, or unsafe `cast`
- "manual" `cast(e,T)`

I tend to think that both should result in the same behavior on static targets, and I think we should have automatic "silent" conversion between numerical types when casting from one to another, because for instance you might have a runtime "Float" (1.5) in JSON typed as Dynamic, assigned to a compile-time "Single".

Here's finally here's for your @Simn

 * Even worse : was about assigning a function var that takes a `Null<Int8>` parameter - hidden behind Dynamic - to a function that takes a `Null<Double>` parameter. For extra fun, both can return different numerical types as well.

**Comment by Simn** [comment]
> Even worse : was about assigning a function var that takes a Null<Int8> parameter - hidden behind Dynamic - to a function that takes a Null<Double> parameter. For extra fun, both can return different numerical types as well.

I still maintain that this violates variance and should not be allowed. It's fine to have this for reference types which can be assumed to have compatible representations between compatible types, but it's not fine for basic types which may even have different sizes. I'm aware that you have funky adapter functions to handle this in HL, but supporting this should not be a requirement in my opinion.

**Comment by RealyUniqueName** [comment]
Yeah, I'd like compiler to not automatically loose anything for me when I use types with strictly defined behavior.

**Comment by Simn** [comment]
JVM issue fixed by #8776. I'll close here, we can argue about casts when we argue about numeric types.
